agent/runner/runner.go

```diff
@@ -0,0 +1,252 @@
+	defaultCapacity      = 32
+)
+
+// Runner executes jobs and actions.
+type Runner struct {
+	l *logrus.Entry
+
+	actions chan actions.Action
+	jobs    chan jobs.Job
+
+	actionsResults chan agentpb.AgentRequestPayload
+	jobsMessages   chan agentpb.AgentResponsePayload
+
+	sem *semaphore.Weighted
+	wg  sync.WaitGroup
+
+	rw      sync.RWMutex
+	rCancel map[string]context.CancelFunc
+}
+
+// New creates new runner. If capacity is 0 then default value is used.
+func New(capacity uint16) *Runner {
+	l := logrus.WithField("component", "runner")
+	if capacity == 0 {
+		capacity = defaultCapacity
+	}
+
+	l.Infof("Runner capacity set to %d.", capacity)
+
+	return &Runner{
+		l:              l,
+		actions:        make(chan actions.Action, bufferSize),
+		jobs:           make(chan jobs.Job, bufferSize),
+		sem:            semaphore.NewWeighted(int64(capacity)),
+		rCancel:        make(map[string]context.CancelFunc),
+		jobsMessages:   make(chan agentpb.AgentResponsePayload),
+		actionsResults: make(chan agentpb.AgentRequestPayload),
+	}
+}
+
+// Run starts jobs execution loop. It reads jobs from the channel and starts them in separate goroutines.
+func (r *Runner) Run(ctx context.Context) {
+	for {
+		select {
+		case action := <-r.actions:
+			r.handleAction(ctx, action)
+		case job := <-r.jobs:
+			r.handleJob(ctx, job)
+		case <-ctx.Done():
+			r.wg.Wait() // wait for all actions and jobs termination
+			close(r.actionsResults)
+			close(r.jobsMessages)
+			return
+		}
+	}
+}
+
+// StartAction starts given actions.Action.
+func (r *Runner) StartAction(action actions.Action) error {
+	select {
+	case r.actions <- action:
+		return nil
+	default:
+		return errors.New("actions queue overflowed")
+	}
+}
+
+// StartJob starts given jobs.Job.
+func (r *Runner) StartJob(job jobs.Job) error {
+	select {
+	case r.jobs <- job:
+		return nil
+	default:
+		return errors.New("actions queue overflowed")
+	}
+}
+
+// JobsMessages returns channel with Jobs messages.
+func (r *Runner) JobsMessages() <-chan agentpb.AgentResponsePayload {
+	return r.jobsMessages
+}
+
+// ActionsResults return chanel with Actions results payload.
+func (r *Runner) ActionsResults() <-chan agentpb.AgentRequestPayload {
+	return r.actionsResults
+}
+
+// Stop stops running Action or Job.
+func (r *Runner) Stop(id string) {
+	r.rw.RLock()
+	defer r.rw.RUnlock()
+
+	// Job removes itself from rCancel map. So here we only invoke cancel.
+	if cancel, ok := r.rCancel[id]; ok {
+		cancel()
+	}
+}
+
+// IsRunning returns true if Action or Job with given ID still running.
+func (r *Runner) IsRunning(id string) bool {
+	r.rw.RLock()
+	defer r.rw.RUnlock()
+	_, ok := r.rCancel[id]
+
+	return ok
+}
+
+func (r *Runner) handleJob(ctx context.Context, job jobs.Job) {
+	jobID, jobType := job.ID(), job.Type()
+	l := r.l.WithFields(logrus.Fields{"id": jobID, "type": jobType})
+
+	if err := r.sem.Acquire(ctx, 1); err != nil {
+		l.Errorf("Failed to acquire token for a job: %v", err)
```

Comment:
Good quiestion!
`Acquire` is blocking operation, if there are no available tokens it will wait. The only one case when this method can return error is when context is terminated, however we can return error message back to the pmm server.

Fixed